### ISSUE TITLE
`-DGCDEBUG=1` is incompatible with `-D_POSIX_C_SOURCE=200112L`

### ISSUE THREAD

**Issue by jpco** [wryun/es-shell#155]
Seems there is some better standards-compliance that can be done in `gc.c`.  I wonder if we could get rid of the big `__MACH__` block while we do.

**Comment by memreflect** [comment]
`__MACH__` was used for any OS based upon the [Mach microkernel](https://en.wikipedia.org/wiki/Mach_(kernel)), which meant mostly NeXT boxes back in the day.
Since the first release of OS X, the XNU kernel for macOS, iOS, Darwin, etc. inherited a lot of NeXT stuff, and the GNU Hurd fork of glibc also can use that code.
That said, both of them can also just use the POSIX code, so i'm fairly certain there's really no need for `__MACH__` anymore.

As for the standards compliance issues unrelated to `__MACH__`, the nonstandard `caddr_t` type [returned by `mmap()` on old systems](https://man.freebsd.org/cgi/man.cgi?query=mmap&apropos=0&sektion=0&manpath=4.4BSD+Lite2&arch=default&format=html) appears to be the only barrier to compiling `gc.c` with `-DGCDEBUG=1 -ansi -D_POSIX_C_SOURCE=200112L on Linux`.
On POSIX systems that support `mmap()`, `mmap()` returns `void *` instead of `caddr_t`.

Speaking of support, `mmap()` is only required for XSI conformance in POSIX.1-2001, meaning it's optional on non-XSI systems, and memory protection with `mprotect()` is also optional and independent of `mmap()`.
In short, a POSIX.1-2001 system could have neither, one, or both of the functions.
POSIX.1-2008 requires both functions even on non-XSI systems.
Either both functions need to be required by a configure test&mdash;if either is missing, configure should fail&mdash;or _es_ needs to assume a POSIX.1-2008 environment at minimum.

Lastly, `MAP_ANONYMOUS` may be named `MAP_ANON`, so it may be worth defining `MAP_ANONYMOUS` in terms of `MAP_ANON` if `MAP_ANONYMOUS` is not already defined.
[POSIX.1-2024 defines both](https://pubs.opengroup.org/onlinepubs/9799919799/basedefs/sys_mman.h.html), so this shouldn't be an issue 5-10 years from now.

**Comment by jpco** [comment]
Sounds like my hope to have a strictly POSIX.1-2001-compliant `GCDEBUG` _es_ is unlikely to happen.  However, simplifying down to what you propose here is probably still valuable for code simplicity in terms of "`#ifdef` considered harmful"
